Server/src/main/java/org/gluu/oxauth/model/common/AuthorizationGrantList.java

```diff
@@ -117,12 +118,23 @@ public ResourceOwnerPasswordCredentialsGrant createResourceOwnerPasswordCredenti
 
     @Override
     public CIBAGrant createCIBAGrant(User user, Client client, int expiresIn) {
+        Long expirationDate = DateUtils.addSeconds(new Date(), expiresIn).getTime();
+        if (appConfiguration.getCibaGrantLifeExtraTimeSec() > 0) {
+            expiresIn += appConfiguration.getCibaGrantLifeExtraTimeSec();
+        }
+
         CIBAGrant grant = grantInstance.select(CIBAGrant.class).get();
         grant.init(user, client, expiresIn);
-
         CIBACacheGrant memcachedGrant = new CIBACacheGrant(grant, appConfiguration);
-        cacheService.put(grant.getCIBAAuthenticationRequestId().getExpiresIn(), memcachedGrant.cacheKey(), memcachedGrant);
+        cacheService.put(grant.getCIBAAuthenticationRequestId().getExpiresIn(),
+                memcachedGrant.cacheKey(), memcachedGrant);
         log.trace("Put CIBA grant in cache, authReqId: " + grant.getCIBAAuthenticationRequestId().getCode() + ", clientId: " + grant.getClientId());
+
+        CIBACacheAuthReqIds cibaCacheAuthReqIds = grantService.getCacheCibaAuthReqIds();
+        cibaCacheAuthReqIds.getAuthReqIds().put(memcachedGrant.cacheKey(), expirationDate);
+        cacheService.put(CIBACacheAuthReqIds.CACHE_KEY, cibaCacheAuthReqIds);
```

Comment:
@miltonbo @qbert2k Would you please explain idea behind it ? It looks like global map of all CIBA auth requests. How is it going to work in cluster? We can't provide thread-safety thus get "lost" requests. 
E.g. if two nodes reads that map and then try to write to it then we will get missed ids:
- node1 reads ["1", "2"] added 3 and trying to write ["1", "2". "3"]
- node2 reads ["1", "2"] added 4 and trying to write ["1", "2". "4"]
Depending on which node writes last you will get missed "3" or "4".

Also we should not call "put" without expiration. I don't think you want to expire all CIBA requests after 60 seconds (which is default if you don't specify expiration for "put"). If you look into `CacheInterface` you will not see such method. It is historic method which was left from ehcache I guess and which we should drop someday. Anyway I marked it as deprecated with this commit https://github.com/GluuFederation/oxCore/commit/8defc9e2e4450ad4a67f02de6aab489139bc18df .

I guess we have similar behavior for "SessionTokens" and "ClientTokens" which are special classes when we force to persist tokens only in cache (for Implicit Flow). Still it's configurable (can be turned off) and it's not global for whole server as it seems in CIBA case.

Reply:
Ok, thanks for your reply. Yeah, I agree with you. I tried to do something similar than other cases as you said, however I understand that it is not thread-safety for the whole app even. The idea behind this is get `PENDING` grants from Cache, however in `CibaRequestsProcessorJob` we don't have authentication request ids to get them from Cache. I could get CIBA clients from data base, however all `authReqIds` are in cache(they are not saved) I think, I'd need another way to implement it then.

Reply:
Ok, thanks for your reply. Yeah, I agree with you. I tried to do something similar than other cases as you said, however I understand that it is not thread-safety for the whole app even. The idea behind this is get `PENDING` grants from Cache, however in `CibaRequestsProcessorJob` we don't have authentication request ids to get them from Cache. I could get CIBA clients from data base, however all `authReqIds` are in cache(they are not saved) I think, I'd need another way to implement it then.

Reply:
Just save it into persistence and grab by state. I think we already discussed it. Cache is bad here because you put key and then in other part you may not be able query it because you don't know key. Persistence allows to query data (while cache don't).